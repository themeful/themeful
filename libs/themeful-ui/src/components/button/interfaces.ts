export type ButtonKind = 'primary' | 'secondary' | 'danger' | 'selectable' | 'menu'
