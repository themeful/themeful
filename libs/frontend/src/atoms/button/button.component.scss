:host {
  display: block;
}

.button {
  border-radius: 0.25rem;
  border: none;
  padding: 8px 16px;
  min-width: 20px;
  min-height: 20px;
  font-weight: 600;
  display: inline-flex;
  cursor: pointer;
  align-items: center;
  justify-content: center;

  --bg-opacity: 0.5;
  --ggs: 0.5;

  &.active {
    --bg-opacity: 0.75;
  }

  &.primary {
    background-color: rgba(50, 90, 240, var(--bg-opacity));
  }

  &.secondary {
    background-color: rgba(200, 200, 200, var(--bg-opacity));
  }

  &.danger {
    background-color: rgba(255, 100, 100, var(--bg-opacity));
  }

  &:disabled {
    filter: grayscale(100%);
    cursor: default;
  }

  &:focus {
    outline: 0 !important;
    box-shadow: 0 0 0 2px rgba(0, 0, 0, 0.5);
  }

  &:focus:not(:focus-visible) {
    box-shadow: none !important;
  }

  &:hover:not(.active):not(:disabled) {
    filter: brightness(0.75);
  }
}

//   &--icon {
//     padding: 2px;
//     width: 24px;
//     height: 24px;
//   }

//   &--small {
//     padding: 4px 8px;
//     font-weight: 400;
//   }

//   & + .m4-button {
//     margin-left: 4px;
//   }

//   &-group {
//     display: flex;

//     & > * {
//       margin-left: 0 !important;
//       flex-grow: 4;

//       &:not(:first-child) {
//         border-top-left-radius: 0;
//         border-bottom-left-radius: 0;
//       }
//       &:not(:last-child) {
//         border-top-right-radius: 0;
//         border-bottom-right-radius: 0;
//       }
//     }
//   }
// }
