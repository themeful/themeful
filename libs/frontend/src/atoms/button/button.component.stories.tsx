import { h } from '@stencil/core'

export default {
  title: 'Atoms/Buttons',
  args: {
    buttons: [
      { label: 'Primary', args: { type: 'button', kind: 'primary' } },
      { label: 'Secondary', args: { type: 'submit', kind: 'secondary' } },
      { label: 'Danger', args: { type: 'button', kind: 'danger' } },
    ],
  },
}

export const buttons = ({ buttons }): HTMLElement => {
  return (
    <div>
      <h1>Buttons</h1>
      <table style={{ borderSpacing: '1rem' }}>
        <tr>
          <td>
            <h3>default</h3>
          </td>
          <td>
            <h3>active</h3>
          </td>
          <td>
            <h3>disabled</h3>
          </td>
        </tr>
        {buttons.map((button) => {
          return (
            <tr>
              <td>
                <tf-button {...button.args}>{button.label}</tf-button>
              </td>
              <td>
                <tf-button {...button.args} active>
                  {button.label}
                </tf-button>
              </td>
              <td>
                <tf-button {...button.args} disabled>
                  {button.label}
                </tf-button>
              </td>
            </tr>
          )
        })}
      </table>
      <h1>With Icon</h1>
      <table>
        <tr>
          <td>default</td>
          <td>
            <tf-button kind="secondary">
              <tf-icon icon="sync"></tf-icon>Some Text
            </tf-button>
          </td>
          <td>
            <tf-button kind="secondary">
              Some Text<tf-icon icon="sync"></tf-icon>
            </tf-button>
          </td>
          <td>
            <tf-button kind="secondary">
              <tf-icon icon="sync"></tf-icon>
            </tf-button>
          </td>
        </tr>
        <tr>
          <td>small</td>
          <td>
            <tf-button kind="secondary" size="small">
              <tf-icon icon="sync"></tf-icon>Some Text
            </tf-button>
          </td>
          <td>
            <tf-button kind="secondary" size="small">
              Some Text<tf-icon icon="sync"></tf-icon>
            </tf-button>
          </td>
          <td>
            <tf-button kind="secondary" size="small">
              <tf-icon icon="sync"></tf-icon>
            </tf-button>
          </td>
        </tr>
        <tr>
          <td>icon</td>
          <td>
            <tf-button kind="secondary" size="icon">
              <tf-icon icon="sync"></tf-icon>Some Text
            </tf-button>
          </td>
          <td>
            <tf-button kind="secondary" size="icon">
              Some Text<tf-icon icon="sync"></tf-icon>
            </tf-button>
          </td>
          <td>
            <tf-button kind="secondary" size="icon">
              <tf-icon icon="sync"></tf-icon>
            </tf-button>
          </td>
        </tr>
      </table>
    </div>
  )
}
