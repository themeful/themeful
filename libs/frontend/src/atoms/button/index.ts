export * from './button.component'
export * from './interfaces'
