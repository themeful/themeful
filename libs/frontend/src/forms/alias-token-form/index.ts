export * from './alias-token-form.component'
