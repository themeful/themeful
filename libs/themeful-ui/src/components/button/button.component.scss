:host {
  display: inline-block;
}

.button {
  color: var(--tf-font-color);
  border-radius: 0.25rem;
  border: none;
  padding: 0.5rem 1rem;
  min-width: 20px;
  width: 100%;
  min-height: 20px;
  cursor: pointer;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 0.25rem;
  font-weight: 600;
  font-size: 0.875rem;
  line-height: 1rem;

  --bg-opacity: 0.5;
  --ggs: 0.5;

  &.small {
    font-weight: 400;
    font-size: 0.75rem;
    line-height: 0.875rem;
    padding: 0.25rem 0.5rem;
  }

  &.icon {
    font-weight: 400;
    font-size: 0.75rem;
    line-height: 0.875rem;
    padding: 0.125rem;
  }

  &.active {
    --bg-opacity: 0.75;
  }

  &.selectable:not(.active),
  &.secondary {
    background-color: rgba(200, 200, 200, var(--bg-opacity));
  }

  :host-context(.link-active) &.selectable,
  &.selectable.active,
  &.primary {
    background-color: rgba(50, 90, 240, var(--bg-opacity));
  }

  &.danger {
    background-color: rgba(255, 100, 100, var(--bg-opacity));
  }

  &:disabled {
    filter: grayscale(100%);
    cursor: default;
  }

  &:focus {
    outline: 0 !important;
    box-shadow: 0 0 0 2px rgba(0, 0, 0, 0.5);
  }

  &:focus:not(:focus-visible) {
    box-shadow: none !important;
  }

  &:hover:not(.active):not(:disabled) {
    filter: brightness(0.75);
  }
}

:host-context(.button-group) {
  .button {
    border-radius: 0;
  }
}

:host(.button-group--first) {
  .button {
    border-top-left-radius: 0.5rem;
    border-bottom-left-radius: 0.5rem;
  }
}

:host(.button-group--last) {
  .button {
    border-top-right-radius: 0.5rem;
    border-bottom-right-radius: 0.5rem;
  }
}
