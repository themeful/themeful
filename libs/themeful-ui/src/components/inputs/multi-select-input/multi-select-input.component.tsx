import { Component, Event, EventEmitter, h, Host, Method, Prop, State, Watch } from '@stencil/core'

@Component({
  tag: 'tf-multi-select-input',
  styleUrl: 'multi-select-input.component.scss',
  shadow: true,
})
export class MultiSelectInputComponent {
  /** Input type */
  @Prop() type = 'text'

  /** Input label */
  @Prop() label: string

  /** Input suggest items */
  @Prop() items: string[] = []

  /** Required input */
  @Prop() required = false

  /** Min input */
  @Prop() minLength: number

  /** Max input */
  @Prop() maxLength: number

  /** Input value */
  @Prop() value: string[] = []

  /** Input Event */
  @Event({ composed: false }) inputChange: EventEmitter

  @State() touched = false
  @State() changed = false
  @State() valid: boolean
  @State() error = ''
  private element: { direction: string; index: number; value: string }

  /** Validate value */
  @Method()
  public validate(): Promise<boolean> {
    this.touched = true
    return Promise.resolve(this.internalValidation())
  }

  /** Changed value */
  @Method()
  public dirty(): Promise<boolean> {
    return Promise.resolve(this.changed)
  }

  @Watch('value')
  public valueChanged(): void {
    this.changed = true
    this.internalValidation()
    this.touched = true
  }

  private internalValidation = (): boolean => {
    this.error = ''
    if (this.required && this.value.length === 0) {
      this.error = `This value is required`
    } else if ((this.value?.length || 0) < this.minLength) {
      this.error = `The min count is ${this.minLength}`
    } else if (this.value?.length > this.maxLength) {
      this.error = `The max count is ${this.maxLength}`
    }
    this.valid = this.error === ''
    return this.valid
  }

  private allowAdd = (event): void => {
    event.preventDefault()
  }

  private add = (): void => {
    if (this.element.value !== null) {
      this.value = [...this.value, this.element.value]
      this.element.value = null
    }
  }

  private addStart = (index): void => {
    this.element = {
      index,
      direction: 'add',
      value: this.items[index],
    }
    this.items = this.items.filter((_, i) => i !== index)
  }

  private addEnd = (): void => {
    if (this.element.value !== null) {
      this.items = [...this.items, this.element.value]
      this.element.value = null
    }
  }

  private allowRemove = (event): void => {
    event.preventDefault()
  }

  private remove = (): void => {
    if (this.element.value !== null) {
      this.items = [...this.items, this.element.value]
      this.element.value = null
    }
  }

  private removeStart = (index): void => {
    this.element = {
      index,
      direction: 'remove',
      value: this.value[index],
    }
    this.value = this.value.filter((_, i) => i !== index)
  }

  private removeEnd = (): void => {
    if (this.element.value !== null) {
      this.value = [...this.value, this.element.value]
      this.element.value = null
    }
  }

  public render(): HTMLTfMultiSelectInputElement {
    return (
      <Host>
        <label
          class={`multi-select-input${
            this.valid && this.touched ? ' multi-select-input--valid' : ''
          }${!this.valid && this.touched ? ' multi-select-input--error' : ''}`}
        >
          <span class="multi-select-input__label">{this.label}</span>
          <div class="multi-select-input__inputs">
            <div class="multi-select-input__col">
              <h4 class="multi-select-input__list-header">Not Assigned</h4>
              <ul
                {...{
                  class: 'multi-select-input__scroll',
                  onDrop: this.remove,
                  onDragOver: this.allowRemove,
                }}
              >
                {this.items?.map((item, index) => (
                  <li
                    {...{
                      draggable: true,
                      onDragStart: () => {
                        this.addStart(index)
                      },
                      onDragEnd: this.addEnd,
                      key: `${index} ${item}`,
                      class: 'multi-select-input__item',
                    }}
                  >
                    {item}
                  </li>
                ))}
              </ul>
            </div>
            <div class="multi-select-input__col">
              <h4 class="multi-select-input__list-header">Assigned</h4>
              <ul
                {...{
                  class: 'multi-select-input__scroll',
                  onDrop: this.add,
                  onDragOver: this.allowAdd,
                }}
              >
                {this.value?.map((item, index) => (
                  <li
                    {...{
                      draggable: true,
                      onDragStart: () => {
                        this.removeStart(index)
                      },
                      onDragEnd: this.removeEnd,
                      key: `${index} ${item}`,
                      class: 'multi-select-input__item',
                    }}
                  >
                    {item}
                  </li>
                ))}
              </ul>
            </div>
          </div>
          <p class="multi-select-input__hint">{this.error}</p>
        </label>
      </Host>
    )
  }
}
