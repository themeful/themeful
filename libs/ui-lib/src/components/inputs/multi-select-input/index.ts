export * from './multi-select-input.component'
