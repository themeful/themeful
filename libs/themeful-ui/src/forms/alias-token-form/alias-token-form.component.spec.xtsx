import { h } from '@stencil/core'
import { newSpecPage } from '@stencil/core/testing'
import { ButtonComponent } from '../../components/button/button.component'
import { MultiSelectInputComponent } from '../../components/inputs/multi-select-input/multi-select-input.component'
import { AliasTokenFormComponent } from './alias-token-form.component'

describe('AliasTokenFormComponent', () => {
  it('renders', async () => {
    const data = {
      formData: {
        identifier: 'dtActionBg',
        aliasTokens: ['atCardBackground', 'atButtonBackground'],
        fields: { selected: [] },
      },
    }
    const page = await newSpecPage({
      components: [AliasTokenFormComponent, ButtonComponent, MultiSelectInputComponent],
      autoApplyChanges: true,
      template: () => <tf-alias-token-form {...data} />,
    })

    expect(page.root).toMatchSnapshot()
  })
})
