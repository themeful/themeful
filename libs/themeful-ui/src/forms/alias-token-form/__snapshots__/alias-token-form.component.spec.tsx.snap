// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AliasTokenFormComponent renders 1`] = `
<tf-alias-token-form>
  <mock:shadow-root>
    <form class="form">
      <h3>
        Select Alias Tokens
      </h3>
      <tf-multi-select-input>
        <mock:shadow-root>
          <label class="multi-select-input">
            <span class="multi-select-input__label"></span>
            <div class="multi-select-input__inputs">
              <div class="multi-select-input__col">
                <h4 class="multi-select-input__list-header">
                  Not Assigned
                </h4>
                <ul class="multi-select-input__scroll">
                  <li class="multi-select-input__item" draggable="true" title="atCardBackground">
                    Card Background
                  </li>
                  <li class="multi-select-input__item" draggable="true" title="atButtonBackground">
                    Button Background
                  </li>
                </ul>
              </div>
              <div class="multi-select-input__col">
                <h4 class="multi-select-input__list-header">
                  Assigned
                </h4>
                <ul class="multi-select-input__scroll"></ul>
              </div>
            </div>
            <p class="multi-select-input__hint"></p>
          </label>
        </mock:shadow-root>
      </tf-multi-select-input>
      <div class="form__controls">
        <tf-button>
          <mock:shadow-root>
            <button class="button secondary" type="button">
              <slot></slot>
            </button>
          </mock:shadow-root>
          Cancel
        </tf-button>
        <tf-button>
          <mock:shadow-root>
            <button class="button primary" type="submit">
              <slot></slot>
            </button>
          </mock:shadow-root>
          Save
        </tf-button>
      </div>
    </form>
  </mock:shadow-root>
</tf-alias-token-form>
`;
